src/core/utils.js

```diff
@@ -305,3 +297,80 @@ export function render404IfConfigKeyIsFalse(
     return <Component {...props} />;
   };
 }
+
+export function getCompatibleVersions({ addon, clientApp } = {}) {
+  let maxVersion = null;
+  let minVersion = null;
+  if (
+    addon && addon.current_version && addon.current_version.compatibility
+  ) {
+    if (addon.current_version.compatibility[clientApp]) {
+      maxVersion = addon.current_version.compatibility[clientApp].max;
+      minVersion = addon.current_version.compatibility[clientApp].min;
+    } else {
+      log.error(
+        'addon found with no compatibility info for valid clientApp',
+        { addon, clientApp }
+      );
+    }
+  }
+
+  return { maxVersion, minVersion };
+}
+
+export function isCompatibleWithUserAgent({
+  _log = log, maxVersion, minVersion, userAgentInfo,
+} = {}) {
+  // If the userAgent is false there was likely a programming error.
+  if (!userAgentInfo) {
+    throw new Error('userAgentInfo is required');
+  }
+
+  const { browser, os } = userAgentInfo;
+
+  // We need a Firefox browser compatible with add-ons (Firefox for iOS does
+  // not currently support add-ons).
+  if (browser.name === 'Firefox' && os.name === 'iOS') {
+    return { compatible: false, reason: INCOMPATIBLE_FIREFOX_FOR_IOS };
+  }
+
+  if (browser.name === 'Firefox') {
+    // Do version checks, if this add-on has minimum or maximum version
+    // requirements.
+    // The mozilla-version-comparator API is quite strange; a result of
+    // `1` means the first argument is higher in version than the second.
+    if (maxVersion && mozCompare(browser.version, maxVersion) === 1) {
+      return { compatible: false, reason: INCOMPATIBLE_OVER_MAX_VERSION };
+    }
+
+    // A result of `-1` means the second argument is a lower version than the
+    // first.
+    if (minVersion && mozCompare(browser.version, minVersion) === -1) {
+      if (minVersion === '*') {
+        _log.error(oneLine`minVersion of "*" was passed to
+          isCompatibleWithUserAgent(); bad add-on version data`,
+          { browserVersion: browser.version, minVersion }
+        );
+      }
+
+      return { compatible: false, reason: INCOMPATIBLE_UNDER_MIN_VERSION };
+    }
+
+    // If we made it here we're compatible (yay!)
+    return { compatible: true };
```

Comment:
Now that I can see the big picture, I think this should set `reason` to `null` because otherwise this result may cause failures in things that check for an `undefined` value. I know that nothing should be using the `reason` value when `compatible` is `false` but `null` makes it more compatible with the other modes just in case.